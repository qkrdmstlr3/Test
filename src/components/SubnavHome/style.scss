@import '/src/common/style/mixin.scss';
@import '/src/common/style/variables.scss';

.nav {
  @include subnav;
  padding: 0 1rem;
  box-sizing: border-box;
  user-select: none;
}
