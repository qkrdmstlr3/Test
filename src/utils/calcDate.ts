const dayList = ['일', '월', '화', '수', '목', '금', '토'];

export function getDday(endDay: string): string {
  const today = new Date();
  const endDate = new Date(`${endDay}T00:00:00`);
  const dday = Math.ceil(
    (endDate.getTime() - today.getTime()) / (1000 * 3600 * 24)
  );

  if (isNaN(dday)) return '';
  if (dday == 0) return 'dday';
  else if (dday < 0) return `+${-dday}`;
  else return `-${dday}`;
}

export function getToday(): string[] {
  const today = new Date();

  const hour = `${today.getHours() < 10 ? '0' : ''}${today.getHours()}`;
  const minute = `${today.getMinutes() < 10 ? '0' : ''}${today.getMinutes()}`;
  const year = `${today.getFullYear()}`;
  const month = `${today.getMonth() + 1}`;
  const day = `${today.getDate()}`;
  const dow = dayList[today.getDay()];

  return [hour, minute, year, month, day, dow];
}
