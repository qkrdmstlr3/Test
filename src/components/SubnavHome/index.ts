import { ShellHTML, createComponent, RenderType } from 'shell-html';
import styleSheet from './style.scss';
import { getToday } from '@Utils/calcDate';

class SubnavHome extends ShellHTML {
  getTime: NodeJS.Timeout | undefined;

  constructor() {
    super([]);
  }

  connectedCallback(): void {
    this.getTime = setInterval(() => {
      this.getTimeNow();
    }, 1000);
  }

  disconnectedCallback(): void {
    if (this.getTime) {
      clearInterval(this.getTime);
    }
  }

  getTimeNow(): void {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [_, minute] = getToday();
    this.setState(minute);
    //TODO: 하루가 지나면 dday도 다시 계산
  }

  render(): RenderType {
    const [hour, minute, year, month, day, dow] = getToday();

    return {
      html: `
      <nav class="nav">
        <div class="nav__date" data-testid="date">${year} . ${month} . ${day}</div>
        <div class="nav__dow" data-testid="dow">${dow}</div>
        <div class="nav__time" data-testid="time">${hour} : ${minute}</div>
      </nav>
      `,
      css: styleSheet,
    };
  }
}

createComponent('subnav-home', SubnavHome);

export default SubnavHome;
