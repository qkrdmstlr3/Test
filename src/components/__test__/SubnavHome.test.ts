/* dependencies */
import { getByTestId } from '@testing-library/dom';
import '@testing-library/jest-dom/extend-expect';

import SubnavHome from '../SubnavHome';
import { getToday } from '@Utils/calcDate';

let subnav: SubnavHome;
let subnavComponent: HTMLElement;

describe('SubnavHome Component test', () => {
  beforeEach(() => {
    subnav = new SubnavHome();
    if (subnav.shadowRoot?.childNodes[0]) {
      subnavComponent = (subnav.shadowRoot as unknown) as HTMLElement;
    }
  });

  it('check date', () => {
    const [hour, minute, year, month, day, dow] = getToday();

    const dateHTML = getByTestId(subnavComponent, 'date').innerHTML;
    expect(dateHTML).toContain(year);
    expect(dateHTML).toContain(month);
    expect(dateHTML).toContain(day);

    const dowHTML = getByTestId(subnavComponent, 'dow').innerHTML;
    expect(dowHTML).toContain(dow);

    const timeHTML = getByTestId(subnavComponent, 'time').innerHTML;
    expect(timeHTML).toContain(hour);
    expect(timeHTML).toContain(minute);
  });
});
