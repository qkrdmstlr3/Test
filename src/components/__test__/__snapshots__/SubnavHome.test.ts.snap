// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubnavHome Component test snapshot 1`] = `
ShadowRoot {
  Symbol(SameObject caches): Object {
    "childNodes": NodeList [
      <nav
        class="nav"
      />,
      <style>
        [object Object]
      </style>,
    ],
  },
}
`;
